rawfile/bd2fs.py

```diff
@@ -284,55 +288,104 @@ def ControllerExpandVolume(self, request, context):
     @log_grpc_request
     def CreateSnapshot(self, request: csi_pb2.CreateSnapshotRequest, context):
         with VolLock(request.source_volume_id):
-            file = img_file(request.source_volume_id)
-            loop_dev = attach_loop(file)
-            fs = from_device(loop_dev)
-            if not fs:
-                raise UnknownFileSystemError(
-                    device=loop_dev, volume_id=request.source_volume_id
-                )
-            fs.create_snapshot(name=request.name)
-
-            if fs.mountpoint is None:
-                detach_loops(file)
-
-            creation_time_ns = time.time_ns()
-            snapshot_id = f"{request.source_volume_id}/{request.name}"
-            # TODO: detach loopdev when we get seperated loop devices
-            nano = 10**9
-            return csi_pb2.CreateSnapshotResponse(
-                snapshot=csi_pb2.Snapshot(
-                    size_bytes=0,
-                    snapshot_id=snapshot_id,
-                    source_volume_id=request.source_volume_id,
-                    creation_time=Timestamp(
-                        seconds=creation_time_ns // nano, nanos=creation_time_ns % nano
-                    ),
-                    ready_to_use=True,
+            volume_meta = metadata(request.source_volume_id)
+            freezefs = volume_meta.get("freezefs", False)
+            copy_on_write_param = volume_meta.get("copy_on_write", None)
+            copy_on_write = (
+                copy_on_write_param
+                if copy_on_write_param is not None
+                else consts.COW_SUPPORTED
+            )
+
+        def _get_current_snapshot():
+            _current = snapshot_manager.list_snapshots(
+                volume_id=request.source_volume_id, snapshot_name=request.name
+            )
+            if not _current.data:
+                return None
+            return _current.data[0]
+
+        nano = 10**9
+        current = _get_current_snapshot()
+        if not current:
+            self._task_manager.run_task(
```

Comment:
there's a race condition here during the time between letting go of the `VolLock` and re-acquiring it here on the `run_task::snapshot_manager::create_snapshot` ?

Reply:
Here we only need lock while reading meta file, nothing will happen if we create multiple snapshots of a single volume at the same time